opnode/rollup/sync/start.go

```diff
@@ -0,0 +1,157 @@
+// Unsafe Block: The highest L2 block. If the L1 Attributes is ahead of the L1 head, it is assumed to be valid,
+// if not, it walks back until it finds the first L2 block whose L1 Origin is canonical in the L1 chain.
+// Safe Block: The highest L2 block whose sequence window has not changed during a reorg.
+func FindL2Heads(ctx context.Context, start eth.L2BlockRef, seqWindowSize uint64,
+	l1 L1Chain, l2 L2Chain, genesis *rollup.Genesis) (unsafe eth.L2BlockRef, safe eth.L2BlockRef, err error) {
+	reorgDepth := 0
+	var prevL1OriginHash common.Hash
+	// First check if the L1 Origin of the start block is ahead of the current L1 head
+	// If so, we assume that this should be the next unsafe head for the sequencing window
+	// We still need to walk back the safe head because we don't know where the reorg started.
+	l1Head, err := l1.L1HeadBlockRef(ctx)
+	if err != nil {
+		return eth.L2BlockRef{}, eth.L2BlockRef{}, err
+	}
+	l2Ahead := start.L1Origin.Number > l1Head.Number
+	var latest eth.L2BlockRef
+
+	// Walk L2 chain until we find the "latest" L2 block. This the first L2 block whose L1 Origin is canonical.
+	for n := start; ; {
+		// Check if l1Origin is canonical when we get to a new epoch
+		if prevL1OriginHash != n.L1Origin.Hash {
+			if ok, err := isCanonical(ctx, l1, n.L1Origin); err != nil {
+				return eth.L2BlockRef{}, eth.L2BlockRef{}, err
+			} else if ok {
+				latest = n
+				break
+			}
+			prevL1OriginHash = n.L1Origin.Hash
+		}
+		// Don't walk past genesis. If we were at the L2 genesis, but could not find the L1 genesis
+		// pointed to from it, we are on the wrong L1 chain.
+		if n.Hash == genesis.L2.Hash || n.Number == genesis.L2.Number {
+			return eth.L2BlockRef{}, eth.L2BlockRef{}, WrongChainErr
+		}
+		// Pull L2 parent for next iteration
+		n, err = l2.L2BlockRefByHash(ctx, n.ParentHash)
+		if err != nil {
+			return eth.L2BlockRef{}, eth.L2BlockRef{}, fmt.Errorf("failed to fetch L2 block by hash %v: %w", n.ParentHash, err)
+		}
+		reorgDepth++
+		if reorgDepth >= MaxReorgDepth {
+			return eth.L2BlockRef{}, eth.L2BlockRef{}, TooDeepReorgErr
+		}
+	}
+	depth := uint64(1) // SeqWindowSize is a length, but we are counting elements in the window.
+	prevL1OriginHash = latest.L1Origin.Hash
+	// Walk from the latest block back 1 Sequence Window of L1 Origins to determine the safe L2 block.
+	for n := latest; ; {
+		// Advance depth if new origin
+		if n.L1Origin.Hash != prevL1OriginHash {
+			depth++
+			prevL1OriginHash = n.L1Origin.Hash
+		}
+		// Walked sufficiently far
+		if depth == seqWindowSize {
+			if l2Ahead {
+				return start, n, nil
+			} else {
+				return latest, n, nil
+			}
+
+		}
+		// Genesis is always safe.
+		if n.Hash == genesis.L2.Hash || n.Number == genesis.L2.Number {
+			safe = eth.L2BlockRef{Hash: genesis.L2.Hash, Number: genesis.L2.Number, Time: genesis.L2Time, L1Origin: genesis.L1}
+			if l2Ahead {
+				return start, safe, nil
+			} else {
+				return latest, safe, nil
+			}
```

Comment:
This is where it is all coming apart from me.

Two scenarios:
1. L1 origin of `start` is ahead of known L1 head. (`l2Ahead == true`)
2. L1 origin of `start` is behind of known L1 head. (`l2Ahead == false`)

In (1), because of the checks earlier in this file (line 99), we know that all L1 origins in the L2 chain are either ahead of the L1 head (\*), or are canonical L1 blocks.

(\*) They are "plausibly" ahead, meaning that there is an extra check (performed when we learn of new L2 blocks) to ensure that (the hash of) the L2 block's L1 origin is not canonical at a height different than the L2 block epoch in the known L1 chain. (This check should probably be mentionned somewhere in this file btw.)

In (2), because again because of line 99's canonicality check, it should always be the case that `start == latest`!
It's the logical combination of:
- The number of the L1 origin of `start` is below the head of the L1 chain (`l2Ahead == false`).
- All L1 origins in the L2 chain are ahead or canonical (line 99 check).

(NB: the `latest` in this code is defined as the highest L2 block such that its L1 block is canonical)

**Conclusions**

As far as I can tell, the behaviour of this algorithm is as follows:
- It validates `start` as the unsafe head against the presence of a L1 reorg (presence of a non-canonical L1 origin).
- If an L1 reorg occurs, it bails.
- It does correctly find the L2 safe block (i.e. the highest L2 block whose epoch's sequencing window is complete in the known L1 chain). Again, assuming no L1 reorg.

This was a bit of a shock, because my prior going into this code was that it was used to "resync" the L2 safe and unsafe block after a reorg was detected. This means that either:
- my understanding is incorrect, and we need to document this stuff better
- this code is incorrect

This needs more investigation, but a quick look at `state.go` where this is called from:
```
	// New L1 Head is not the same as the current head or a single step linear extension.
	// This could either be a long L1 extension, or a reorg. Both can be handled the same way.
	s.log.Warn("L1 Head signal indicates an L1 re-org", "old_l1_head", s.l1Head, "new_l1_head_parent", newL1Head.ParentHash, "new_l1_head", newL1Head)
	unsafeL2Head, safeL2Head, err := sync.FindL2Heads(ctx, s.l2Head, s.Config.SeqWindowSize, s.l1, s.l2, &s.Config.Genesis)
	if err != nil {
		s.log.Error("Could not get new unsafe L2 head when trying to handle a re-org", "err", err)
		return err
	}
```

It looks like a L1 reorg (assuming we've built L2 blocks on some of the L1 block re-orged away) will always result in an error here. Of course, using the old L2 head, we will encounter L1 origins that are non canonical since L1 re-orged!

Reply:
Followup question to sanity check all of this: do we test L1 reorgs anywhere? Doesn't seem to be the case in `start_test.go`.